pkg/ccl/importccl/exportcsv.go

```diff
@@ -0,0 +1,320 @@
+				return err
+			}
+		}
+
+		filename := exportFilePatternDefault
+		if override, ok := opts[exportOptionChunkSize]; ok {
+			filename = override
+		}
+		if strings.Count(filename, exportFilePatternPart) != 1 {
+			return errors.Errorf("EXPORT filename pattern does must contain one instance of %q", exportFilePatternPart)
+		}
+
+		out := distsqlrun.ProcessorCoreUnion{CSVWriter: &distsqlrun.CSVWriterSpec{
+			Destination: file,
+			NamePattern: filename,
+			Delimiter:   delim,
+			ChunkRows:   int64(chunk),
+		}}
+
+		rows := sqlbase.NewRowContainer(
+			p.ExtendedEvalContext().Mon.MakeBoundAccount(), sqlbase.ColTypeInfoFromColTypes(sql.ExportPlanResultTypes), 0,
+		)
+		rw := sql.NewRowResultWriter(rows)
+
+		if err := p.DistSQLPlanner().PlanAndRunExport(
+			ctx, p.Txn(), p.ExtendedEvalContext(), plans[0], out, rw,
+		); err != nil {
+			return err
+		}
+		for i := 0; i < rows.Len(); i++ {
+			resultsCh <- rows.At(i)
+		}
+		rows.Close(ctx)
+		return rw.Err()
+	}
+
+	return fn, exportHeader, []sql.PlanNode{sel}, nil
+}
+
+func newCSVWriterProcessor(
+	flowCtx *distsqlrun.FlowCtx,
+	spec distsqlrun.CSVWriterSpec,
+	input distsqlrun.RowSource,
+	output distsqlrun.RowReceiver,
+) (distsqlrun.Processor, error) {
+	c := &csvWriter{
+		flowCtx:  flowCtx,
+		spec:     spec,
+		input:    input,
+		output:   output,
+		settings: flowCtx.Settings,
+		registry: flowCtx.JobRegistry,
+		progress: spec.Progress,
+		db:       flowCtx.EvalCtx.Txn.DB(),
+	}
+	if err := c.out.Init(&distsqlrun.PostProcessSpec{}, sql.ExportPlanResultTypes, flowCtx.NewEvalCtx(), output); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type csvWriter struct {
+	flowCtx  *distsqlrun.FlowCtx
+	spec     distsqlrun.CSVWriterSpec
+	input    distsqlrun.RowSource
+	out      distsqlrun.ProcOutputHelper
+	output   distsqlrun.RowReceiver
+	settings *cluster.Settings
+	registry *jobs.Registry
+	progress distsqlrun.JobProgress
+	db       *client.DB
+}
+
+var _ distsqlrun.Processor = &csvWriter{}
+
+func (sp *csvWriter) OutputTypes() []sqlbase.ColumnType {
+	return sql.ExportPlanResultTypes
+}
+
+func (sp *csvWriter) Run(ctx context.Context, wg *sync.WaitGroup) {
+	ctx, span := tracing.ChildSpan(ctx, "csvWriter")
+	defer tracing.FinishSpan(span)
+
+	if wg != nil {
+		defer wg.Done()
+	}
+
+	err := func() error {
+		pattern := exportFilePatternDefault
+		if sp.spec.NamePattern != "" {
+			pattern = sp.spec.NamePattern
+		}
+
+		types := sp.input.OutputTypes()
+		sp.input.Start(ctx)
+		input := distsqlrun.MakeNoMetadataRowSource(sp.input, sp.output)
+
+		alloc := &sqlbase.DatumAlloc{}
+
+		var buf bytes.Buffer
+		writer := csv.NewWriter(&buf)
+		if sp.spec.Delimiter != 0 {
+			writer.Comma = sp.spec.Delimiter
+		}
+
+		f := tree.NewFmtCtxWithBuf(tree.FmtParsable)
+		defer f.Close()
+
+		csvRow := make([]string, len(types))
+
+		chunk := 0
+		done := false
+		var rows int64
+		for {
+			buf.Reset()
+			for {
+				if sp.spec.ChunkRows > 0 && rows >= sp.spec.ChunkRows {
```

Comment:
Yeah, when we added retry to our uploads, we lost streaming since we don't want to obligate the producer be able to re-produce arbitrary amounts.

Maybe we just add a default chunk size of e.g. 10000.

Reply:
I was thinking buf.Len so we care about bytes more than row count. But whatever.